### PR TITLE
Cache Watch

### PR DESCRIPTION
**Pull request by banks**
**Note:** this is against `f-envoy` as it depends on other cache changes made in this branch.

Add Watch method to agent cache that abstracts local blocking on the cache.

### Why

Cache was designed so that it mimics our blocking interface to consumers while managing the actual blocking loop to RPC separately. This makes sense and makes it very easy to work with when using it to cache RPC responses for use in normal blocking HTTP endpoints.

For Proxy config I need to implement other internal processes that "watch" for changes through the cache and while it's possible to rewrite the blocking loop logic in a bunch of other places, it gets messy and repetitive to do so and test them all for edge cases like backoff on error etc.

It was also hard to generalise that logic outside of the cache package (it would be easier with the added Index meta here but still this seems generic enough that it warrants being first class in the cache package API rather than re-invented elsewhere).

### Concerns

Firstly, although the blocking loop is conceptually simple, this cost me another several hours of trying to reason about how to build this which is unfortunate. One of the reasons I wanted to build it once and for all into the cache package is so we don't have to think about it again in other packages that need internal "watches" on results.

I find it kinda gross though that this is now the 3rd separate blocking loop in play in the cache package. When they are all used together it is _super_ hard to reason about what is happening - blocking in Watch, blocking in Get, potentially blocking in refresh etc.

In my immediate use-case background refresh is not actually that useful and this could replace it.

Currently, background refresh is either on or off for a whole cache type which means users of this library will have all 3 blocking loops interacting.

At some later date I feel like it might be cleaner to replace background refresh with something like this:
 - any component that needs to ensure a cached result is available would create a goroutine that watches using `Watch` and reacts as needed to changes. It's free to let that goroutine live as long as the state is necessary.
 - other consumers of the same RPC/cache type that don't necessarily need background refresh don't have to have it
 - the cache will still de-dupe the actual RPCs if there are multiple callers all blocking on the same thing
 - taken like this, Watch can be seen as an "on demand" background refresh that is both easier to consume from an internal component and more flexible.

But this is on the critical path for some work so I don't want to make that more invasive refactor at this stage.

There is a final point which is that it would be possible to get rid of this additional blocking loop in a couple of other ways but they all seem to have other tradeoffs:
 - Augment the cache-types themselves that you care about so that they report updates to an external component as well as the cache and then rely on background refresh to keep that triggering. This seems like a bad pattern to me.
 - refactor `Get` into something more generic so that it's "blocking loop" (i.e. run through to `<-waiterCh` and then `goto RETRY_GET`) can be re-used both for a single response and for delivering updates on a chan. This could simplify things a bit but its more work that I want to put in immediately.

### CODE REVIEW

File: agent/cache/cache.go [hunk 1]
```diff
@@ -230,6 +231,11 @@ RETRY_GET:
 		return nil, ResultMeta{}, fmt.Errorf("unknown type in cache: %s", t)
 	}
 
+	// Get the current value
+	c.entriesLock.RLock()
+	entry, ok := c.entries[key]
+	c.entriesLock.RUnlock()
+
```
Comment (banks) [inline agent/cache/cache.go]: Moved this as I realised I was shadowing `ok` with my previous change to load `tEntry` in a previous PR. It didn't break any tests presumably because `entry.Valid` is always `false` on a zero-value `Entry` anyway and it the second `ok` was false we would have errored above. But still seemed less jank to not have it in the wrong order!

File: agent/cache/watch_test.go [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+			require.Error(u.Err)
+		}
+	}
+	// Must be fewer than 10 failures in that time
+	require.True(numErrors < 10, fmt.Sprintf("numErrors: %d", numErrors))
+
+	// Check the number of RPCs as a sanity check too
+	actual := atomic.LoadUint32(&retries)
+	require.True(actual < 10, fmt.Sprintf("actual: %d", actual))
+}
```
Comment (banks) [inline agent/cache/watch_test.go]: Note: I did have another test here to test badly-behaved RPCs that return zero values, but it turns out that it only exercised the existing code in `Get` and failed for a subtle reason - the test was both returning 0 index AND never blocking even if the request came in with `index = 1`. This is not behaviour we expect of any RPC that uses blockingQuery helper _even_ if it's badly behaved and returns zero index. Fixing `Get` to handle taht would be a ton of refactoring for a case that can never realistically exist.

Stuff like this is a good example of why so many layers of blocking is hard to work with!

File: agent/cache/watch.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	info.MinIndex = r.minIndex
+	info.Timeout = 10 * time.Minute
+	return info
+}
+
+// Watch behaves like Get but return a chan to be watched for updates
+// asynchronously. It is a helper that abstracts code from perfroming their own
+// "blocking" query logic against a cache key to watch for changes and to
+// maintain the key in cache actively. It will continue to perform blocking Get
+// requests until the context is canceled.
```
Comment (mitchellh) [inline agent/cache/watch.go]: Probably worth documenting that the watcher will block if the channel write blocks, so care should be taken by the reader to read quickly.

File: agent/cache/watch.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+			// Check context hasn't been cancelled
+			if ctx.Err() != nil {
+				return
+			}
+
+			// Check the index of the value returned in the cache entry to be sure it
+			// changed
+			if wrappedR.minIndex < meta.Index {
+				u := WatchUpdate{res, meta, err}
+				ch <- u
```
Comment (mitchellh) [inline agent/cache/watch.go]: I would wrap this in a `select` on `ctx.Done()` too so that a cancellation doesn't keep the goroutine running forever if the other side doesn't drain.

File: agent/cache/watch.go (hunk 3 above)
Comment (banks) [reply agent/cache/watch.go]: Good catch.

File: agent/cache/watch.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+	Err    error
+}
+
+// indexManagedRequest is a wrapper around a user-passed Request that allows us
+// to manage the blocking index without knowing how to modify the concrete
+// Request type.
+type indexManagedRequest struct {
+	r        Request
+	minIndex uint64
+}
```
Comment (banks) [inline agent/cache/watch.go]: @mitchellh FYI, this turns out to be a big problem with this design...

All the tests with mock cache types work wonderfully, but when I'd got all the way to using this for real somewhere, the real cache types all type-check the request object and error because they get an `indexManagedRequest` and not the type they wanted 😢 .

I'll re-think how we can abstract managing the query options on request types without wrapping the request object I guess.

File: agent/cache/watch.go (hunk 3 above)
Comment (banks) [reply agent/cache/watch.go]: Added a comment (incoming) but it turns out this is probably never a problem - worst that can happen is that no cache Get's are made for long enough for the result to be evicted, but even then as soon as the caller drains the chan again, new fetching will resume...

Most important bad behaviour to document it not leaving the watch without ever cancelling it as that leaks the goroutine.

**Comment by banks** [comment]
@mitchellh sorry, but I changed the API to make multiple service discovery watches more efficient in the upstream code. There wasn't much point supporting to slightly different APIs for this so I just replaced `Watch` here with `Notify`.

File: agent/cache/watch.go [hunk 6]
```diff
@@ -6,51 +6,67 @@ import (
...
+		return fmt.Errorf("watch requires the type to support blocking")
 	}
 
 	// Always start at 0 index to deliver the inital (possibly currently cached
 	// value).
 	index := uint64(0)
 
 	go func() {
-		defer close(ch)
+		defer fmt.Println("Exited", correlationID)
```
Comment (mitchellh) [inline agent/cache/watch.go]: Probably don't want to be printing to stdout from deep in a lib. :)

Comment (banks) [reply agent/cache/watch.go]: Hmmm I removed this in my last commit: https://github.com/hashicorp/consul/pull/4695/commits/c2be69354ea792beb548e5ae6793080375f5e39a

Your comment even shows up as "outdated" for me. Wonder how that happened.

Thanks though - I saw that after I pushed the change.

**Comment by banks** [comment]
@mitchellh your concern was valid but somehow it was already fixed when you looked - maybe github UI gave you an old changes link?

This is rebased on `f-envoy` since I merged the other 2 PRs, any chance you can give it a final thumbs up?

File: agent/cache/watch.go [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+	// choose to assign random IDs for example.
+	CorrelationID string
+	Result        interface{}
+	Meta          ResultMeta
+	Err           error
+}
+
+// Notify registers a desire to be updated about changes to a cache result.
+//
+// It is a helper that abstracts code from perfroming their own "blocking" query
```
Comment (rboyer) [inline agent/cache/watch.go]: typo in comment: `s/perfroming/performing/`

File: agent/cache/watch.go [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+	tEntry, ok := c.types[t]
+	c.typesLock.RUnlock()
+	if !ok {
+		return fmt.Errorf("unknown type in cache: %s", t)
+	}
+	if !tEntry.Type.SupportsBlocking() {
+		return fmt.Errorf("watch requires the type to support blocking")
+	}
+
+	// Always start at 0 index to deliver the inital (possibly currently cached
```
Comment (rboyer) [inline agent/cache/watch.go]: typo: `s/inital/initial/`

File: agent/cache/watch.go [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+		var failures uint
+
+		for {
+			// Check context hasn't been cancelled
+			if ctx.Err() != nil {
+				return
+			}
+
+			// Blocking request
+			res, meta, err := c.getWithIndex(t, r, index)
```
Comment (rboyer) [inline agent/cache/watch.go]: Would it make sense to plumb `ctx` down into `getWithIndex` so the blocking query can terminate quickly when the context is cancelled?

Comment (mitchellh) [reply agent/cache/watch.go]: I would love this, and I'll let Banks respond since he has more context (😏) but if I recall our internal RPC client doesn't play nice with contexts so this would require plumbing _all the way down_. I took a look at this when I first wrote the cache package. Its something I'd love to see but would be a high touch change and probably warrants its own big PR.

Additionally, since these functions are all unexported, it makes it easy[easier] to do this in the future since it doesn't break user API. :)
